pre-commit/notebook_check_ast.py

```diff
@@ -16,7 +16,7 @@ def main(files: List[Path]) -> bool:
         for n, cell in enumerate(notebook["cells"]):
             if cell["cell_type"] != "code":
                 continue
-            source = "".join([x for x in cell["source"] if not x.startswith("!")])
+            source = "".join([x for x in cell["source"] if x[0] not in ("#", "%", "!")])
```

Comment:
This is causing it to fail on Windows.

Reply:
Can you give any more detail? Like the error output? I cannot see any reason why this should behave any differently on a different platform. I need more information.

Reply:
`
check notebook ast.......................................................Failed
- hook id: notebook-check-ast
- exit code: 1
Traceback (most recent call last):
  File "pre-commit/notebook_check_ast.py", line 33, in <module>
    sys.exit(1 - main(args.files))
  File "pre-commit/notebook_check_ast.py", line 15, in main
    notebook = json.load(fh)
  File "c:\tools\anaconda3\envs\tiatoolbox_dev\lib\json\__init__.py", line 293, in load
    return loads(fp.read(),
  File "c:\tools\anaconda3\envs\tiatoolbox_dev\lib\encodings\cp1252.py", line 23, in decode
    return codecs.charmap_decode(input,self.errors,decoding_table)[0]
UnicodeDecodeError: 'charmap' codec can't decode byte 0x9d in position 4641: character maps to <undefined>
notebook urls............................................................Failed
- hook id: notebook-urls
- exit code: 1
examples\01-wsi-reading.ipynb
Traceback (most recent call last):
  File "pre-commit/notebook_urls.py", line 93, in <module>
    sys.exit(1 - main(args.files))
  File "pre-commit/notebook_urls.py", line 55, in main
    notebook = json.load(fh)
  File "c:\tools\anaconda3\envs\tiatoolbox_dev\lib\json\__init__.py", line 293, in load
    return loads(fp.read(),
  File "c:\tools\anaconda3\envs\tiatoolbox_dev\lib\encodings\cp1252.py", line 23, in decode
    return codecs.charmap_decode(input,self.errors,decoding_table)[0]
UnicodeDecodeError: 'charmap' codec can't decode byte 0x9d in position 4641: character maps to <undefined>
notebook markdown format.................................................Failed
- hook id: notebook-markdown-format
- exit code: 1
Traceback (most recent call last):
  File "pre-commit/notebook_markdown_format.py", line 74, in <module>
    main(sorted(args.files))
  File "pre-commit/notebook_markdown_format.py", line 52, in main
    notebook = json.loads(path.read_text())
  File "c:\tools\anaconda3\envs\tiatoolbox_dev\lib\pathlib.py", line 1222, in read_text
    return f.read()
  File "c:\tools\anaconda3\envs\tiatoolbox_dev\lib\encodings\cp1252.py", line 23, in decode
    return codecs.charmap_decode(input,self.errors,decoding_table)[0]
UnicodeDecodeError: 'charmap' codec can't decode byte 0x9d in position 4641: character maps to <undefined>
`

Reply:
Ok thank you, this traceback indicates the issue is not on this line. It is a file encoding issue. It appears windows is not using utf-8 for some reason.

Reply:
See https://github.com/TissueImageAnalytics/tiatoolbox/pull/519/commits/f9de1fa39a457921343cc18ae22300ccdb5e6c92